turbo/txpool-provider/block_diff.go

```diff
@@ -0,0 +1,187 @@
+package txpoolprovider
+
+import (
+	"math"
+	"math/big"
+
+	"github.com/ledgerwatch/turbo-geth/common"
+	"github.com/ledgerwatch/turbo-geth/core/rawdb"
+	"github.com/ledgerwatch/turbo-geth/core/types"
+	"github.com/ledgerwatch/turbo-geth/core/types/accounts"
+	"github.com/ledgerwatch/turbo-geth/ethdb"
+	"github.com/ledgerwatch/turbo-geth/rlp"
+	"github.com/ledgerwatch/turbo-geth/turbo/adapter"
+	"github.com/ledgerwatch/turbo-geth/turbo/txpool-provider/pb"
+)
+
+// buildBlockDiff is a helper function which constructs a block diff to send to
+// subscribers.
+//
+// There are three different scenarios that must be handled:
+//
+// 1) The subscriber has requested to receive the latest block information. This
+//    is done by setting newHead to nil. The block diff will be generated
+//    against the head and its parent. This will always generate an
+//    AppliedBlock type of diff.
+// 2) The chain head has moved forward one block. In this case, the newest block
+//    was applied to the last head.
+// 3) The chain head has changed and its parent is not equal to the old head. In
+//    that case, the fork should be analyzed and all changes should be sent to
+//    the subscriber.
+func buildBlockDiff(oldHead, newHead *types.Header, chainId *big.Int, db ethdb.Database) (*pb.BlockDiff, error) {
```

Comment:
Can this function accept the pairs (block_hash; block_height) instead of the entire block header? I can see that only hash and number is used. If so, then you can skip couple of `ReadHeaderByHash` below when you already know hash and block height?

Reply:
Actually, the parent hash of the new head is used to determine if the block is being applied to current canonical chain or if it was unwound. Do you still prefer to pass all those fields individually?